js/private/js_run_devserver.js

```diff
@@ -0,0 +1,165 @@
+                `Syncing symlink ${src.slice(RUNFILES_ROOT.length + 1)}`
+            )
+            await mkdirp(path.dirname(dst))
+            if (exists) {
+                await fs.promises.unlink(dst)
+            }
+            await fs.promises.symlink(src, dst)
+        } else if (lstat.isDirectory()) {
+            await mkdirp(dst)
+            const contents = await fs.promises.readdir(src)
+            await Promise.all(
+                contents.map((entry) =>
+                    syncRecursive(path.join(src, entry), path.join(dst, entry))
+                )
+            )
+        } else {
+            console.log(`Syncing file ${src.slice(RUNFILES_ROOT.length + 1)}`)
+            await mkdirp(path.dirname(dst))
+            if (exists) {
+                await fs.promises.unlink(dst)
+            }
+            await fs.promises.copyFile(src, dst)
+            const mode = isExecutable(lstat) ? 0o744 : 0x644
+            await fs.promises.chmod(dst, mode)
+        }
+    } catch (e) {
+        console.error(e)
+        process.exit(1)
+    }
+}
+
+async function sync(files, rootdir, sync = false) {
+    for (const file of files) {
+        const src = path.join(RUNFILES_ROOT, file)
+        const dst = path.join(rootdir, file)
+        if (sync) {
+            await syncRecursive(src, dst)
+        } else {
+            syncRecursive(src, dst)
+        }
+    }
+}
+
+async function main(args, rootdir) {
+    console.log(
+        `\n\nStarting js_run_devserver ${process.env.JS_BINARY__TARGET}`
+    )
+
+    const configPath = path.join(RUNFILES_ROOT, args[0])
+
+    const config = JSON.parse(await fs.promises.readFile(configPath))
+
+    await sync(config.data_files, rootdir, true)
+    console.log('Initial sync done')
+
+    return new Promise((resolve) => {
+        const cwd = process.env.JS_BINARY__CHDIR
+            ? path.join(rootdir, process.env.JS_BINARY__CHDIR)
+            : rootdir
+
+        const tool = config.tool
+            ? path.join(RUNFILES_ROOT, config.tool)
+            : config.command
+
+        const toolArgs = args.slice(1)
+
+        console.log(`Running '${tool} ${toolArgs.join(' ')}' in ${cwd}\n\n`)
+
+        const proc = child_process.spawn(tool, toolArgs, {
+            cwd: cwd,
+            stdio: 'inherit',
+            env: {
+                ...process.env,
+                BAZEL_BINDIR: '.',
+                JS_BINARY__CHDIR: '',
+            },
+        })
+
+        proc.on('close', (code) => {
+            console.error(`child tool process exited with code ${code}`)
+            resolve()
+            process.exit(code)
+        })
+
+        process.stdin.on('data', (chunk) => {
+            const chunkString = chunk.toString()
+            if (chunkString.indexOf('IBAZEL_BUILD_COMPLETED SUCCESS') !== -1) {
+                console.log('IBAZEL_BUILD_COMPLETED SUCCESS')
+                // Re-parse the config file to get the latest list of data files to copy
+                sync(
```

Comment:
Since we aren't doing `await`... what if another ibazel `IBAZEL_BUILD_COMPLETED` comes in while we are in the middle of a sync? Then we'll have 2 concurrent syncs that might conflict?

Reply:
Can't we just add `async` here so each `IBAZEL_BUILD_COMPLETED` waits for the previous to complete?

Reply:
Yeah. That is probably best and simples. And then figure out how to do any many parallel promises as possible in sync itself.

Reply:
Actually if I'm recalling the `data_files` correctly I don't think we even need recursion and can have one giant `Promise.all`?
```
function sync(...) { return Promise.all(files.map(... syncRecursive(src, dst)) }

function syncRecursive(src, dst) {
   ...if/lstat/check-cache...

   await mkdirp(path.dirname(dst));
   if (symlink) ... link ...
   else if (file) ... copy/chmod ...
   else if (directory) ... mkdir() ...    // <--- no "p"
}
```

Note that the results of `syncRecursive` are put in `Promise.all` and not individual `await`s so it should be 100% parallel?